test/compatibility/dogstatsd_datagram_compatibility_test.rb

```diff
@@ -75,28 +108,32 @@ def with_legacy_client
     ensure
       new_backend.socket.close if new_backend&.socket
       StatsD.backend = old_backend
+      StatsD.prefix = old_prefix
+      StatsD.default_tags = old_default_tags
     end
 
-    def with_new_client
+    def with_new_client(mode)
+      prefix = mode == :with_prefix ? 'prefix' : nil
+      default_tags = mode == :with_default_tags ? { key: 'value' } : nil
       client = StatsD::Instrument::Client.new(
         sink: StatsD::Instrument::UDPSink.new(@host, @port),
         datagram_builder_class: StatsD::Instrument::DogStatsDDatagramBuilder,
+        prefix: prefix,
+        default_tags: default_tags
       )
 
       yield(client)
     end
 
-    def assert_equal_datagrams_emitted(&block)
-      legacy_datagram = with_legacy_client { |client| read_datagram(client, &block) }
-      new_datagram = with_new_client { |client| read_datagram(client, &block) }
-
-      assert_equal legacy_datagram, new_datagram, "The datagrams emitted from both clients were not the same"
-    end
-
     def drain_server
-      loop { @server.recvfrom_nonblock(100) }
-    rescue
-      IO::EAGAINWaitReadable
+      retried = false
+      loop do
+        @server.recvfrom_nonblock(100)
+        retried = false
+      end
+    rescue IO::WaitReadable
+      retried = true
+      retry unless retried
```

Comment:
Hum, this is fishy. You set `retried = true` just above, so you'll never `retry` AFAICT.

Also when `IO::EAGAINWaitReadable` is raised you're suppose to wait with `IO.select`